core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -200,17 +217,33 @@ public void run() {
           }
 
           ResolutionResults resolutionResults;
-          try {
-            ResourceResolver resourceResolver = null;
-            if (enableJndi) {
-              resourceResolver = getResourceResolver();
+          long networkAddressCacheTtl = getNetworkAddressCacheTtl();
+          boolean refreshRequired = networkAddressCacheTtl <= 0
+              || cachedResolutionResults == null
+              || stopwatch.elapsed(TimeUnit.SECONDS) >= networkAddressCacheTtl;
+
+          if (refreshRequired) {
+            try {
+              ResourceResolver resourceResolver = null;
+              if (enableJndi) {
+                resourceResolver = getResourceResolver();
+              }
+
+              resolutionResults =
+                  resolveAll(addressResolver, resourceResolver, enableSrv, enableTxt, host);
+              if (networkAddressCacheTtl > 0) {
+                cachedResolutionResults = resolutionResults;
+                stopwatch.reset().start();
+              }
+            } catch (Exception e) {
+              savedListener.onError(
+                  Status.UNAVAILABLE
+                      .withDescription("Unable to resolve host " + host)
+                      .withCause(e));
+              return;
             }
-            resolutionResults =
-                resolveAll(addressResolver, resourceResolver, enableSrv, enableTxt, host);
-          } catch (Exception e) {
-            savedListener.onError(
-                Status.UNAVAILABLE.withDescription("Unable to resolve host " + host).withCause(e));
-            return;
+          } else {
+            resolutionResults = cachedResolutionResults;
```

Comment:
It seems like in this case we should just return now. Since the values we pass to `onAddresses()` will be the same, there's no need to call it.